silhouette/app/com/mohiva/play/silhouette/impl/providers/oauth2/Auth0Provider.scala

```diff
@@ -0,0 +1,198 @@
+              throw new ProfileRetrievalException(s"Cannot get user profile - HTTP status code ${response.status}")
+          }
+        }
+      }
+    }
+  }
+}
+
+case class Auth0SocialProfile(loginInfo: LoginInfo,
+  nickname: String,
+  email: Option[String],
+  emailVerified: Option[Boolean],
+  avatarURL: Option[String] = None,
+  firstName: Option[String] = None,
+  lastName: Option[String] = None,
+  fullName: Option[String] = None,
+  identities: Seq[LoginInfo]) extends IdentifiedSocialProfile
+/**
+ * The profile parser for the common social profile.
+ */
+class Auth0ProfileParser extends SocialProfileParser[JsValue, Auth0SocialProfile, OAuth2Info] {
+
+  /**
+   * Parses the social profile.
+   *
+   * @param json The content returned from the provider.
+   * @return The social profile from given result.
+   */
+  override def parse(json: JsValue, authInfo: OAuth2Info): Future[Auth0SocialProfile] = Future.successful {
+    val userID = (json \ "user_id").as[String]
+    val avatarURL = (json \ "picture").asOpt[String]
+    val email = (json \ "email").asOpt[String]
+    val emailVerified = (json \ "email_verified").asOpt[Boolean]
+    val nickname = (json \ "nickname").as[String]
+    val identities = (json \ "identities").as[Seq[JsValue]] map {
+      jsIdentity =>
+        {
+          val provider = (jsIdentity \ "provider").as[String]
+          val providerUserId = (jsIdentity \ "user_id").as[String]
+          LoginInfo(provider, providerUserId)
+        }
+    }
+
+    Auth0SocialProfile(
+      loginInfo = LoginInfo(ID, userID),
+      nickname = nickname,
+      avatarURL = avatarURL,
+      email = email,
+      emailVerified = emailVerified,
+      identities = identities)
+  }
+}
+
+/**
+ * The Auth0 OAuth2 Provider.
+ *
+ * @param httpLayer The HTTP layer implementation.
+ * @param stateProvider The state provider implementation.
+ * @param settings The provider settings.
+ */
+class Auth0Provider(
+  protected val httpLayer: HTTPLayer,
+  protected val stateProvider: OAuth2StateProvider,
+  val settings: OAuth2Settings)
+  extends BaseAuth0Provider with Auth0ProfileBuilder {
+
+  /**
+   * The type of this class.
+   */
+  override type Self = Auth0Provider
+
+  /**
+   * The profile parser implementation.
+   */
+  override val profileParser = new Auth0ProfileParser
+
+  /**
+   * Gets a provider initialized with a new settings object.
+   *
+   * @param f A function which gets the settings passed and returns different settings.
+   * @return An instance of the provider initialized with new settings.
+   */
+  override def withSettings(f: (Settings) => Settings) = new Auth0Provider(httpLayer, stateProvider, f(settings))
+
+  /**
+   * Gets the access token.
+   *
+   * @param code    The access code.
+   * @param request The current request.
+   * @return The info containing the access token.
+   */
+  override protected def getAccessToken(code: String)(implicit request: RequestHeader): Future[OAuth2Info] = {
```

Comment:
This method should be placed in the `BaseAuth0Provider` so that it can be reused.